src/core/dev/allocator.cpp

```diff
@@ -363,3 +379,195 @@ bool xlio_allocator_hw::register_memory(ib_ctx_handler *p_ib_ctx_h)
 {
     return m_data && xlio_registrator::register_memory(m_data, m_size, p_ib_ctx_h);
 }
+
+/*
+ * xlio_allocator_heap implementation
+ */
+
+struct heap_key {
+    alloc_t alloc_func;
+    free_t free_func;
+    bool hw;
+
+    bool operator==(const heap_key &key) const
+    {
+        return key.alloc_func == alloc_func && key.free_func == free_func && key.hw == hw;
+    }
+};
+
+namespace std {
+template <> class hash<heap_key> {
+public:
+    size_t operator()(const heap_key &key) const
+    {
+        return (((size_t)key.alloc_func + (size_t)key.free_func) & ~(size_t)1LU) | key.hw;
+    }
+};
+} // namespace std
+
+#define XLIO_HEAP_METADATA_BLOCK (16U * 1024U * 1024U)
+
+static std::unordered_map<heap_key, xlio_heap *> s_heap_map;
+static lock_mutex s_heap_lock;
+static size_t s_pagesize;
+
+/*static*/
+xlio_heap *xlio_heap::get(alloc_t alloc_func, free_t free_func, bool hw)
+{
+    std::lock_guard<decltype(s_heap_lock)> lock(s_heap_lock);
+
+    if (!alloc_func) {
+        // Free is pointless without allocation function. Reset it for the key.
+        free_func = nullptr;
+    }
+
+    heap_key key = {.alloc_func = alloc_func, .free_func = free_func, .hw = hw};
+    auto item = s_heap_map.find(key);
+    xlio_heap *heap = (item == s_heap_map.end()) ? nullptr : item->second;
+
+    if (!heap) {
+        heap = new xlio_heap(alloc_func, free_func, hw);
+        s_heap_map[key] = heap;
+    }
+    return heap;
+}
+
+/*static*/
+void xlio_heap::initialize()
+{
+    // Cache pagesize to align allocation requests.
+    s_pagesize = (size_t)sysconf(_SC_PAGESIZE) ?: 4096U;
```

Comment:
must be a power of 2, right? should we check it?

Reply:
you're right, it must be power of 2 to make a mask from it. According to manual page, this sysconf() call returns pagesize and cannot fail. So it's unexpected to even to get 0 here. We rely on power 2 in other places too. So if we want to add such checks, they should be added in a single place and then all the users should use this single obtained and checked value.

So I'd prefer not to add a check here to avoid complications.